sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -521,38 +524,138 @@ class Analyzer(
+      resolvedAttrs.filterNot(attr => outputSet.exists(_.semanticEquals(attr)))
+    }
+
+    private def findNotResolvedMissingAttrs(
+        outputSet: Seq[Expression],
+        inputSet: AttributeSet,
+        missingAttrs: Seq[Attribute]): Seq[Attribute] = {
+      val resolvedAttrs =
+        missingAttrs.filter(attr => inputSet.exists(_.semanticEquals(attr)))
+      resolvedAttrs.filterNot(attr => outputSet.exists(_.semanticEquals(attr)))
+    }
+
+    /**
+     * Traverse the tree until resolving the sorting attributes
+     * Return all the resolvable missing sorting attributes
+     */
+    @tailrec
+    private def collectResolvableMissingAttrs(
+        ordering: Seq[SortOrder],
+        plans: mutable.Queue[LogicalPlan]): (Seq[SortOrder], Seq[Attribute]) = {
+      if (plans.isEmpty) (Seq.empty[SortOrder], Seq.empty[Attribute])
+      else {
+        plans.dequeue() match {
+          // Only Windows and Project have projectList-like attribute.
+          case un: UnaryNode if un.isInstanceOf[Project] || un.isInstanceOf[Window] =>
+            val (newOrdering, missingAttrs) = resolveAndFindMissing(ordering, un, un.child)
+            // If missingAttrs is non empty, that means we got it and return it;
+            // Otherwise, continue to traverse the tree.
+            if (missingAttrs.nonEmpty) (newOrdering, missingAttrs)
+            else {
+              plans.enqueue(un.child)
+              collectResolvableMissingAttrs(ordering, plans)
+            }
+          // Jump over the following UnaryNode types
+          // The output of these types is the same as their child's output
+          case un: UnaryNode
+            if un.isInstanceOf[Distinct] ||
+              un.isInstanceOf[Filter] ||
+              un.isInstanceOf[Limit] ||
+              un.isInstanceOf[RedistributeData] ||
+              un.isInstanceOf[Repartition] ||
+              un.isInstanceOf[RepartitionByExpression] ||
+              un.isInstanceOf[Sample] ||
+              un.isInstanceOf[Sort] ||
+              un.isInstanceOf[SortPartitions] ||
+              un.isInstanceOf[Subquery] ||
+              un.isInstanceOf[With] ||
+              un.isInstanceOf[WithWindowDefinition] =>
+            assert(un.inputSet == un.outputSet)
+            plans.enqueue(un.child)
+            collectResolvableMissingAttrs(ordering, plans)
+          case a: Aggregate =>
+            val (newOrdering, missingAttrs) = resolveAndFindMissing(ordering, a, a.child)
+            // For Aggregate, all the order by columns must be specified in group by clauses
+            if (missingAttrs.nonEmpty &&
+                missingAttrs.forall(ar => a.groupingExpressions.exists(_.semanticEquals(ar)))) {
+              (newOrdering, missingAttrs)
+            }
+            // If missingAttrs is empty, do not traverse its child and just try the next
+            else collectResolvableMissingAttrs(ordering, plans)
+          case join @ Join(left, right, joinType, _) =>
```

Comment:
I think we should stop at Join, if it can't be resolved by outputSet of Join, then we can't resolve it, because the output of Join should already include all the outputSet of the children (exclude semi join). The projection/aggregation below Join are part of subquery, we could not resolve the ordering from subqueries.